### ISSUE TITLE
Rewrite dd.read_hdf

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1408]
HDF format is probably the most mature cross-platform format in common use today for pandas-like data.  Dask.dataframe supports this format through `dd.to_hdf` and `dd.read_hdf` functions, which are decently full featured.  

However, as various features have glommed onto `read_hdf` it has become fairly complex, causing significant maintenance burden and addition friction when adding new features.  As we think about naturally supporting intelligent partitioning (see #1407), a new and important feature, we might also consider how `read_hdf` might be rewritten from scratch to improve things.

It would be nice to collect both some current frustrations and some long-term desires for dealing with dask.dataframe and HDF files together so that future designs can have these in mind. 

cc @nirizr @alberthdev

**Comment by mrocklin** [comment]
In regards to cleanly supporting partitions I think that a design like the following might be helpful: 

```python
If `*` is somewhere in the inputs:
    filenames, datapaths = find_all_matching_datapaths(filename, datapath)
    filenames, datapaths = sort_by_name(filenames, datapaths)
    dfs = [dd.read_hdf(fn, datapath) for fn, datapath in zip(filenames, datapaths)]
    if not mostly_sorted_divisions(dfs):
        dfs = [df.clear_divisions() for df in dfs]
    return dd.concat(dfs)  # use logic in concat to realign slightly overlapping partitions
else:
    return old_single_datapath_version_of_read_hdf(filename, datapath)
```

Pieces that are yet to be implemented include `find_all_matching_datapaths` which, as brought up by @alberthdev can be quite slow if done naively and `mostly_sorted_divisions`, which, I think, should be fairly straightforward.

**Comment by nirizr** [comment]
I think this also relates to #1058

**Comment by mrocklin** [comment]
Pinging @thequackdaddy

**Comment by mrocklin** [comment]
Comment from @nirizr in https://github.com/dask/dask/pull/1407#issuecomment-234782556

> Regarding your ideal read_hdf pseudo-code, I think the separation between "healing a broken df" (as discussed in #1058) should be kept here.

Yeah, I was thinking about that when writing up that pseudocode. I think that dd.concat is exactly that function, especially with the reliance on the dd.multi.align_partitions function, which performs much of the magic that we need here.

**Comment by nirizr** [comment]
Continuing the conversation there, whouldn't you say the following code (from your original pseudo-code) should be part of the healing process?
```
    if not mostly_sorted_divisions(dfs):
        dfs = [df.clear_divisions() for df in dfs]
```

I would also try keeping both handling a single datapath in the same flow if possible, perhaps we could rewrite without completely separating the two. That's probably not too high on the wishlist, though.

**Comment by nirizr** [comment]
We might want to support fixed format now, as mentioned in #1167.
It is now possible to save 'fixed' globstring dataframes and then load them as multiple partitions. This should be in line with dasks preference for multiple partitions.

**Comment by nirizr** [comment]
Another radical thought: If reading a slice out of a single hdf dataset is not frequently used, it might be simpler to just drop `start` and `stop` to keep `read_hdf` simple. Users still wishing to get a slice could use dask operations instead.
This will also break pandas API compatibility.
I'm not too sure this will be worth it and how much usage `start` and `stop` has, but I think it's worth bringing up.

**Comment by mrocklin** [comment]
> I would also try keeping both handling a single datapath in the same flow if possible, perhaps we could rewrite without completely separating the two. That's probably not too high on the wishlist, though.

Agreed.  This would work if we read a single file in chunks by making many single-partition dask.dataframes and then called `concat` on all of them as we do now for the `*` case.

> We might want to support fixed format now, as mentioned in #1167.
> It is now possible to save 'fixed' globstring dataframes and then load them as multiple partitions. This should be in line with dasks preference for multiple partitions

Agreed that fixed with `*` now makes sense.  Good idea.

> Another radical thought: If reading a slice out of a single hdf dataset is not frequently used, it might be simpler to just drop start and stop to keep read_hdf simple. Users still wishing to get a slice could use dask operations instead.

I'm curious if anyone uses this.  I wouldn't be terribly surprised if a few people did although I agree that it's lower priority.

**Comment by nirizr** [comment]
format='table' now raises a warning.
We might want to consider removing the (just added) warning if we're contemplating supporting format=fixed.

**Comment by nirizr** [comment]
@mrocklin you might want to consider my last comment and removing the format parameter warning before it gets to users.

**Comment by mrocklin** [comment]
Unfortunately this comes roughly 30 seconds too late :)

Releases are cheap.  We can do another soon.

**Comment by nirizr** [comment]
Oh well, I'll try being more alert next time :D

**Comment by mrocklin** [comment]
And I should really try to be more transparent / public about release
cycles.

**Comment by thequackdaddy** [comment]
Just an update... for my purposes, I've really switched to dask/bcolz. My tables are very columnar, so column-oriented tables are much better for me. 

So I really haven't played with HDFStores much recently. Happy to provide

**Comment by alberthdev** [comment]
I don't think we use slicing much... though keeping sane with the API might be useful, and those who are more used to Pandas may appreciate this feature quite a bit! That said, if lazy loading is implemented in `dd.read_hdf` (I assume it is), I think using the Dask's built-in slicing should work as well. (The slice should hopefully be loaded on demand!)

Another side thought - `dd.read_hdf` (or some alternative API) should also be able to support raw HDF5 reading, e.g. a la H5py style reading. There are a lot of HDF5 files out there that do not use the Pandas format for storing data, and it would be immensely useful to have dask support these kind of raw files.

**Comment by nirizr** [comment]
Trying to summarize requests/suggestions so far:

1. simplify currently over-complex code
2. support reading fixed format
3. remove slice (`start`, `stop`) functionality, point users to dask's builtin slicing tools.

**Comment by mrocklin** [comment]
My pseudocode above fails to properly divide single large files:

```python
If `*` is somewhere in the inputs:
    filenames, datapaths = find_all_matching_datapaths(filename, datapath)
    filenames, datapaths = sort_by_name(filenames, datapaths)
    dfs = [dd.read_hdf(fn, datapath) for fn, datapath in zip(filenames, datapaths)]
    if not mostly_sorted_divisions(dfs):
        dfs = [df.clear_divisions() for df in dfs]
    return dd.concat(dfs)  # use logic in concat to realign slightly overlapping partitions
else:
    return old_single_datapath_version_of_read_hdf(filename, datapath)
```

Because we still have the problem of "how do I find rows to cleanly divide up this single file"

This could be resolved if instead of these lines:

    filenames, datapaths = find_all_matching_datapaths(filename, datapath)
    filenames, datapaths = sort_by_name(filenames, datapaths)
    dfs = [dd.read_hdf(fn, datapath) for fn, datapath in zip(filenames, datapaths)]


We had these lines

    filenames, datapaths, kwargs_list = find_all_pd_read_hdf_calls(filename, datapath)
    filenames, datapaths, kwargs_list = sort_by_name(filenames, datapaths, kwargs_list)
    dfs = [dd.read_hdf(fn, datapath, **kwargs) for fn, datapath, kwargs in zip(filenames, datapaths, kwargs_list)]

Where now `kwargs_list` might optionally include start/stop entries for the `pd.read_hdf` call.  That is we make a new dask.dataframe for *every* partition, even for partitions within a single file/datapath.

**Comment by mrocklin** [comment]
A further advantage is that most of the hard logic is now probably separated from reading files, so should be way easier to test and, I think easier to maintain and extend.

**Comment by ankravch** [comment]
Hi there, Is there plan for dd.read_hdf to support 'where', 'columns' keywords by exploiting what's originally done in pd.read_hdf?

**Comment by mrocklin** [comment]
The API of pandas is generally in scope.  It depends on who shows up to implement these features.
